st2common/st2common/util/config_loader.py

```diff
@@ -112,9 +114,56 @@ def _get_object_property_schema(object_schema, additional_properties_keys=None):
             # ensure that these keys are present in the object
             for key in additional_properties_keys:
                 property_schema[key] = additional_properties
-        property_schema.update(object_schema.get("properties", {}))
+
+        properties_schema = object_schema.get("properties", {})
+        property_schema.update(properties_schema)
+
+        potential_patterned_keys = set(additional_properties_keys) - set(
+            properties_schema.keys()
+        )
+
+        pattern_properties = object_schema.get("patternProperties", {})
+        # patternProperties can be a boolean or a dict
+        if pattern_properties and isinstance(pattern_properties, dict):
+            # update any matching key
+            for raw_pattern, pattern_schema in pattern_properties.items():
+                if not potential_patterned_keys:
+                    # nothing to check. Don't compile any more patterns
+                    break
+                pattern = re.compile(raw_pattern)
+                for key in list(potential_patterned_keys):
+                    if pattern.search(key):
+                        property_schema[key] = pattern_schema
+                        potential_patterned_keys.remove(key)
```

Comment:
Can we add into the tests a case where we have additionalProperties as a list and a patternProperties. Show that the additionalProperty that doesn't match the pattern has its schema applied, but where the additionalProperty matches the patternProperty that the patternProperty one is applied.
Is there a need to remove the key from the potential_patterned_keys set whilst we are traversing it? It's just a temporary set.

Reply:
Why would additionalProperties be a list? Is this yet another schema feature I didn't account for?

Reply:
We remove the key from the set so that we don't match a key more than once. The first found pattern wins.

Reply:
I was going by this: http://json-schema.org/understanding-json-schema/reference/object.html#additional-properties - where it has an example of additionalproperties that is a list.

Reply:
As discussed on Slack, correction from me - additioanProperties is object or boolean.

Reply:
Ok. I've pushed my refactor - the properties logic feels a lot cleaner now. Next chance I get I'll add tests that show the order of precedence of properties, patternProperties, and additionalProperties.